File: src/engine.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
核心算法引擎模块 (src/engine.py)
规则：
1. 绝对不跨标点符号断句切片。
2. N字严格替换N字（无声调拼音逐字完全相同）。
"""

import re
from pypinyin import Style, lazy_pinyin


class HomophonicEngine:
    def __init__(self, dictionary_words):
        self.pinyin_map = {}
        for word in set(dictionary_words):
            clean_word = word.strip()
            if not clean_word:
                continue
            p_full = tuple(lazy_pinyin(clean_word, style=Style.TONE))
            p_norm = tuple(lazy_pinyin(clean_word, style=Style.NORMAL))

            item = {
                "word": clean_word,
                "length": len(clean_word),
                "pinyin_full": p_full,
                "pinyin_norm": p_norm,
            }
            if p_norm not in self.pinyin_map:
                self.pinyin_map[p_norm] = []
            self.pinyin_map[p_norm].append(item)

    def _split_into_subsentences(self, text):
        sub_parts = re.split(r"([，。；？！、\n\r\t“”《》兮])", text)
        result = []
        curr_offset = 0
        for part in sub_parts:
            if not part or re.search(r"[，。；？！、\n\r\t“”《》兮]", part):
                curr_offset += len(part)
                continue
            chars = [ch for ch in part if "\u4e00" <= ch <= "\u9fa5"]
            indices = [
                curr_offset + idx
                for idx, ch in enumerate(part)
                if "\u4e00" <= ch <= "\u9fa5"
            ]
            if chars:
                result.append({"sub_text": "".join(chars), "indices": indices})
            curr_offset += len(part)
        return result

    def find_puns(self, full_sentence, min_len=2, max_len=4):
        matches = []
        sub_sents = self._split_into_subsentences(full_sentence)
        for sub in sub_sents:
            chars = sub["sub_text"]
            indices = sub["indices"]
            n = len(chars)

            for length in range(min_len, max_len + 1):
                if length > n:
                    continue
                for i in range(n - length + 1):
                    sub_chars = chars[i : i + length]
                    sub_p_full = tuple(lazy_pinyin(sub_chars, style=Style.TONE))
                    sub_p_norm = tuple(
                        lazy_pinyin(sub_chars, style=Style.NORMAL)
                    )

                    if sub_p_norm in self.pinyin_map:
                        for item in self.pinyin_map[sub_p_norm]:
                            if sub_chars == item["word"]:
                                continue
                            is_same_tone = sub_p_full == item["pinyin_full"]
                            match_type = (
                                "全同音同调" if is_same_tone else "全同音异声调"
                            )
                            start_idx = indices[i]
                            end_idx = indices[i + length - 1] + 1
                            pun_sent = (
                                full_sentence[:start_idx]
                                + f"【{item['word']}】"
                                + full_sentence[end_idx:]
                            )

                            matches.append(
                                {
                                    "original_text": sub_chars,
                                    "replaced_word": item["word"],
                                    "length": length,
                                    "match_type": match_type,
                                    "is_same_tone": is_same_tone,
                                    "pun_sentence": pun_sent,
                                    "pinyin_orig": " ".join(sub_p_full),
                                    "pinyin_target": " ".join(
                                        item["pinyin_full"]
                                    ),
                                }
                            )
        seen = set()
        return [
            m
            for m in matches
            if not (
                (m["pun_sentence"], m["replaced_word"]) in seen
                or seen.add((m["pun_sentence"], m["replaced_word"]))
            )
        ]

File: src/engine.js
/**
 * Homophonic matching engine.
 *
 * Rules:
 * 1. Never create a window across punctuation or paragraph boundaries.
 * 2. Match N Han characters against N Han characters.
 * 3. Resolve polyphonic characters from the complete clause, then slice the
 *    resolved pinyin arrays. Never re-pronounce an isolated 2–4 character span.
 */

const { pinyin, customPinyin } = require('pinyin-pro');

const HAN_RE = /\p{Script=Han}/u;
const BOUNDARY_RE = /[，,。．.;；:：?？!！、…—–\-\n\r\t“”‘’「」『』《》〈〉（）()【】\[\]{}]/u;
const PINYIN_OPTIONS = { type: 'array', toneSandhi: false, segmentit: 2 };

class HomophonicEngine {
  constructor(dictionaryWords, options = {}) {
    this.pinyinMap = new Map();
    if (options.pinyinOverrides && Object.keys(options.pinyinOverrides).length) {
      customPinyin(options.pinyinOverrides, { polyphonic: 'replace' });
    }

    const uniqueWords = Array.from(new Set(dictionaryWords));
    for (const rawWord of uniqueWords) {
      const cleanWord = rawWord.trim();
      if (!cleanWord) continue;
      const pFull = this._toPinyin(cleanWord, 'symbol').join(' ');
      const pNormKey = this._toPinyin(cleanWord, 'none').join(' ');
      const item = {
        word: cleanWord,
        length: Array.from(cleanWord).length,
        pinyinFull: pFull,
        pinyinNormKey: pNormKey
      };
      if (!this.pinyinMap.has(pNormKey)) this.pinyinMap.set(pNormKey, []);
      this.pinyinMap.get(pNormKey).push(item);
    }
  }

  _toPinyin(text, toneType) {
    return pinyin(text, { ...PINYIN_OPTIONS, toneType });
  }

  _splitIntoSubsentences(text) {
    const result = [];
    let chars = [];
    let indices = [];

    const flush = () => {
      if (chars.length) result.push({ subText: chars.join(''), indices });
      chars = [];
      indices = [];
    };

    let codeUnitOffset = 0;
    for (const character of text) {
      if (BOUNDARY_RE.test(character)) {
        flush();
      } else if (HAN_RE.test(character)) {
        chars.push(character);
        indices.push(codeUnitOffset);
      }
      codeUnitOffset += character.length;
    }
    flush();
    return result;
  }

  findPuns(fullSentence, minLen = 2, maxLen = 4) {
    const matches = [];
    const subSents = this._splitIntoSubsentences(fullSentence);

    for (const sub of subSents) {
      const chars = sub.subText;
      const indices = sub.indices;
      const n = Array.from(chars).length;
      const contextFull = this._toPinyin(chars, 'symbol');
      const contextNorm = this._toPinyin(chars, 'none');
      if (contextFull.length !== n || contextNorm.length !== n) {
        throw new Error(`拼音与汉字未对齐：${chars}`);
      }

      for (let length = minLen; length <= maxLen; length++) {
        if (length > n) continue;
        for (let i = 0; i <= n - length; i++) {
          const subChars = Array.from(chars).slice(i, i + length).join('');
          const subPFull = contextFull.slice(i, i + length).join(' ');
          const subPNormKey = contextNorm.slice(i, i + length).join(' ');
          const candidateItems = this.pinyinMap.get(subPNormKey) || [];

          for (const item of candidateItems) {
            if (subChars === item.word) continue;
            const isSameTone = subPFull === item.pinyinFull;
            const startIdx = indices[i];
            const endIdx = indices[i + length - 1] + 1;
            matches.push({
              original_text: subChars,
              replaced_word: item.word,
              length,
              match_type: isSameTone ? '全同音同调' : '全同音异声调',
              is_same_tone: isSameTone,
              pun_sentence: `${fullSentence.slice(0, startIdx)}【${item.word}】${fullSentence.slice(endIdx)}`,
              pinyin_orig: subPFull,
              pinyin_target: item.pinyinFull
            });
          }
        }
      }
    }

    const seen = new Set();
    return matches.filter(match => {
      const key = `${match.pun_sentence}_${match.replaced_word}`;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }
}

module.exports = HomophonicEngine;
module.exports.BOUNDARY_RE = BOUNDARY_RE;
